### ISSUE TITLE
Steam Controller not working

### ISSUE THREAD

**Issue by aboutafter** [libretro/Lakka#279]
The Steam Controller doesn't work in Lakka. I'm not sure if it's possible to make it work, but it would be a very nice addition if possible. 

The problem I see with it is that without Steam, I think it only works as a mouse, so the majority of keys don't do anything (or at least it doesn't recognize almost anything when I try to bind it manually). Once Steam is open you can change modes (it imitates an Xbox 360 gamepad, or WASD keys and other templates), so maybe there is a mode in Lakka to make it behave like a Xbox 360 gamepad?

**Comment by Kivutar** [comment]
I found this https://github.com/libretro/RetroArch/issues/2390

Have you tried to do a manual bind?

**Comment by aboutafter** [comment]
Hi Kivutar, yes I did, but the only key that works it's the "A" key.

**Comment by aboutafter** [comment]
Maybe Lakka needs something like this project so the controller could work easily? https://github.com/ynsta/steamcontroller

**Comment by johnjelinek** [comment]
unfortunately, the driver above requires python -- which doesn't ship w/ Lakka

**Comment by johnjelinek** [comment]
I'm not sure if this driver might work: https://github.com/cvuchener/steamcontroller-linux-kernel I don't know how to install it in lakka.

**Comment by Kivutar** [comment]
I saw some kernel patches about the steam controller in openelec

**Comment by Kivutar** [comment]
Or maybe just this? http://askubuntu.com/questions/686214/how-do-i-get-a-steam-controller-working

**Comment by aboutafter** [comment]
No, with that it only works in basic mode.

**Comment by Kivutar** [comment]
What's the difference between basic mode and other modes?

**Comment by johnjelinek** [comment]
It's hard to explain without showing you. It's unique in that the touchpads
double as trackpads (eg: mouse emulation) or as analog sticks or buttons.
Are you suggesting to jump into the terminal in lakka to get uinput working?

**Comment by Kivutar** [comment]
I saw some udev rules to enable mouse/trackpad emulation too (through uinput yeah).

Is this the way to go?

Lakka (OpenELEC) allows you to add custom udev rules in the storage partition. We could use that way to experiment.

**Comment by aboutafter** [comment]
There is news about the driver I posted in January: 

https://www.gamingonlinux.com/articles/theres-an-open-source-standalone-steam-controller-driver.7015

Apparently it already works in Xbox 360 controller mode. Is there any possibility to add this to Lakka or the fact that it requires python is a no-no as @johnjelinek says?

**Comment by Kivutar** [comment]
Yeah the python part is a no-no, sorry. (I don't know how people nowadays think that adding python as dependancy for a joypad driver can be smart)

**Comment by johnjelinek** [comment]
Well, Python is included in Debian and Ubuntu out of the box, so most
people prolly don't think they're taking on an extra dependency, but I know
what you mean. Do you think all userland drivers should be discarded and
only kernel modules should be adopted?

**Comment by Kivutar** [comment]
For very important controllers like this one, we can do exceptions and have a userspace driver. In an ideal world, we would keep all this stuff kernel side.

Anyway, the real blocker here is that I don't own the Steam Controller. If one of you own the device, can you try adding the support for it and send a pull request please?

**Comment by johnjelinek** [comment]
I own it, but the Python driver gets a little buggy sometimes where it
thinks there are multiple Xbox controllers plugged in. There's another
driver that is a kernel module, but it has different issues. For most
games, it works really well though.
On Apr 13, 2016 9:51 AM, "Jean-André Santoni" <notifications@github.com>
wrote:

For very important controllers like this one, we can do exceptions and have
a userspace driver. In an ideal world, we would keep all this stuff kernel
side.

Anyway, the real blocker here is that I don't own the Steam Controller. If
one of you own the device, can you try adding the support for it and send a
pull request please?

—
You are receiving this because you were mentioned.
Reply to this email directly or view it on GitHub
<https://github.com/libretro/Lakka/issues/279#issuecomment-209486845>

**Comment by aboutafter** [comment]
How about this one I've just saw on reddit:

https://www.reddit.com/r/linux_gaming/comments/4eneek/steam_controller_daemon/
https://gitlab.com/dennis-hamester/scd

**Comment by johnjelinek** [comment]
I'll have to try that one out, what's the deploy story for a c++ userland
driver?

**Comment by Kivutar** [comment]
I just received my steam controller. I got it working in basic keyboard+mouse mode, with steam launched. I'm going to try the various userland drivers now.

**Comment by Kivutar** [comment]
I plugged the controller dongle, paired it with the central button, then launched Steam.

I launched RetroArch, saw a yellow message saying that an XBOX 360 controller was recognized.

I did a Bind All, and was able to map some of the keys:
 - the directions to the analog pad
 - ABXY to BAYX
 - start and select
 - L and R

I couldn't bind any of the L2 and R2 and L3 and R3. No right analog either. And the directional cross of the steam controller was unresponsive too.

That was enough to play SNES games.

**Comment by tolkonepiu** [comment]
https://github.com/LibreELEC/LibreELEC.tv/blob/master/packages/addons/driver/steamcontroller/package.mk
